### ISSUE TITLE
@workspace query hangs indefinitely on "Collecting workspace info" step

### ISSUE THREAD

**Issue by zpincus** [microsoft/vscode-copilot-release#676]
- VS Code Version: 1.85.1
- OS Version: macOS 14.2

Steps to Reproduce:
1. Make a workspace chat request in the chat sidebar (not inline), e.g. "@workspace Hey is this thing working?"
2. Wait indefinitely while the chat process hangs at the "collecting workspace info" step:
<img width="350" alt="image" src="https://github.com/microsoft/vscode-copilot-release/assets/174217/a5c7aa95-2788-42b4-829f-d61dff8a390d">

GitHub Copilot Chat output from clicking "send" on a chat request:
```2023-12-14 12:52:44.850 [info] [chat fetch] engine https://api.githubcopilot.com/chat
2023-12-14 12:52:44.850 [info] [chat fetch] modelMaxTokenWindow 4096
2023-12-14 12:52:44.850 [info] [chat fetch] chat model gpt-4
2023-12-14 12:52:50.825 [info] [chat fetch] request.response: [https://api.githubcopilot.com/chat/completions], took 5974 ms
2023-12-14 12:52:50.903 [info] [streamMessages] message 0 returned. finish reason: [stop]
2023-12-14 12:52:50.904 [info] [streamChoices] request done: requestId: [26c6c167-820f-46c9-a629-239adba7c0af] responseId: [79fc550d-4eb8-46e4-8b1c-1691e678171e] model deployment ID: [x74a4b0e7ace1]
```

The workspace is a large git repo with _many_ files, should that be relevant. I'm not sure what the next steps are to debug / diagnose. There appears to be no other relevant information in any other log files, except for perhaps the `Window` log which reports cycling unresponsiveness of the extension host:
```
2023-12-14 12:53:18.932 [info] Extension host (LocalProcess pid: 44483) is unresponsive.
2023-12-14 12:53:21.525 [info] Extension host (LocalProcess pid: 44483) is responsive.
2023-12-14 12:53:29.622 [info] Extension host (LocalProcess pid: 44483) is unresponsive.
2023-12-14 12:53:30.974 [info] Extension host (LocalProcess pid: 44483) is responsive.
2023-12-14 12:57:19.036 [info] Extension host (LocalProcess pid: 44483) is unresponsive.
2023-12-14 12:57:22.906 [info] Extension host (LocalProcess pid: 44483) is responsive.
2023-12-14 12:57:33.821 [info] Extension host (LocalProcess pid: 44483) is unresponsive.
2023-12-14 12:57:40.907 [info] Extension host (LocalProcess pid: 44483) is responsive.
```

Regular copilot completions and inline chat both seem to work fine.

**Comment by liuxiang951023** [comment]
在吗

**Comment by andreifoldes** [comment]
I noticed this too. My repo is not large, but it does contain a data folder with 0.5TB which could be causing the problem? Is there a way to tell copilot to ignore it during indexing?

**Comment by zpincus** [comment]
For the record, I can get the copilot chat working using my exact same extensions and so forth in a window not connected to an open directory, or with a window opened to my home directory or similar, or with a variety of other medium-sized repos opened.

But it seems to hang forever collecting information on my larger repo, which contains moderately more code and a lot more ancillary and data files. The whole repo is ~14 GB, and contains about 20 MB of python files all told, ~7 GiB of jupyter notebook files, and much of the rest in assorted text and tsv files containing various data.

Also, this issue applies to not just `@workspace` queries, but it seems to any chat requests (except inline chat).

**Comment by mjbvz** [comment]
Yes that's the problem. Copilot skips binary files but needs to index text files. If there are multi GB of text, this is going to be slow

Can you try creating a `.copilotignore` file at the roof of your workspace that excludes the data folders? This file works pretty much just like a `.gitignore`

**Comment by zpincus** [comment]
Great -- excluding the many GB of data and .ipynb files fixes things right up.

I was not aware of `.copilotignore`, nor of the possibility that it would even try to index that much non-code text from the repo; thank you! Is there a whitelist (`.copilotinclude` or whatever) approach too that can be used? E.g. to just allow code files or specific relevant data files?

**Comment by andreifoldes** [comment]
Relatedly, is the cache of the indexing stored locally? Should one leave the "Collecting workspace info" run as long as it needs in hopes of faster incremental indexing happening at later requests?

**Comment by DarshanRamakant** [comment]
I am also facing the similar issue. In my case I am running it on the LLVM repo, which is relatively large. But I am not sure which files to exclude here ( there are no binaries in my workspace)

**Comment by PhKP** [comment]
I'm also facing this issue when using `@workspace`

I'm using a large monorepo which is about 30 GB in size with 200.000 files and 80.000 folders. Out of these the .git folder is 10 GB with about 3500 files and 3000 folders. I can understand why these numbers would be hard for copilot to work with. I tried to implement a `.copilotignore` file, but I'm in doubt if it works as I think it does (like gitignore). Without any documentation it's hard to understand if it's a real feature which sets hard boundaries or it's just a file that copilot happens to look into when "_deciding what workspace information to collect_". In any case, copilot seems to not hang indefinitely any more, but doesn't really seem to know about things I know isn't ignored in the  `.copilotignore` file either.

**Comment by Yehuda-Bergstein** [comment]
.copilotignore doesn't work for me... did it work for anybody else?

**Comment by kaonmir** [comment]
Well, Content exclusions function might have been rollback [[article](https://github.blog/changelog/2023-11-20-copilot-content-exclusions-temporary-rollback-and-upcoming-fix/)]

But I am not sure this is related to `.copilotignore`...

**Comment by WiktorrWieSolteq** [comment]
Any updates guys?

**Comment by DomagojKorais** [comment]
Any update? My copilot is still at `Gathering workspace info`.

**Comment by Junosprite007** [comment]
I'm going on 2 ish hours of `Gathering workspace info`, but I guess my workspace is pretty big with lots of unrelated projects. I should just leave my computer running then?

Is there a way to get a progress bar or something on this process or nah? I mean I guess it'll only need to happen once per workspace, so only once for me, Ideally.

**Comment by adityaraj-r** [comment]
I'm facing the same issue. It keeps saying `Gathering workspace info` and then stops without giving any response
I tried `.copilotignore` solution but it's not working even now. 
I am working on a repo inside a docker container using Remote Explorer extension in VS Code 1.86.2 on Mac OS 14.4

**Comment by ipstone** [comment]
what is the convention to use to specify the folder, type of files to exclude from @workspace index? - similar to .gitignore ?

Another related question is whether we can specify a subfolder to index the code base?

much thanks!
